File: Programing/HigherOrderFunction/Exercise2/AnswerJava/Exercise2.java
import java.util.*;
import java.util.stream.*;

class Exercise2 {
    public void execute() {
	Calculator calculator = new Calculator();
	// 演算方法を登録
	calculator.addOperator(Operator.Add, this::add);
	calculator.addOperator(Operator.Sub, this::subtract);
	calculator.addOperator(Operator.Mul, this::multiply);
	calculator.addOperator(Operator.Div, this::divide);

	// 演算方法を指定して演算
	System.out.println(calculator.calculate(Operator.Add, 12, 30)); // 出力 : 42
	System.out.println(calculator.calculate(Operator.Div, 252, 6)); // 出力 : 42
    }

    private Integer add(Integer a, Integer b) { return a + b; }
    private Integer subtract(Integer a, Integer b) { return a - b; }
    private Integer multiply(Integer a, Integer b) { return a * b; }
    private Integer divide(Integer a, Integer b) { return a / b; }
}

File: Programing/HigherOrderFunction/Exercise2/AnswerCSharp/Exercise2.cs
using System;

namespace Exercise2
{
    class Exercise2
    {
        public void Execute()
	{
	    var calculator = new Calculator();
	    // 演算方法を登録
	    calculator.AddOperator(Operator.Add, add);
	    calculator.AddOperator(Operator.Sub, subtract);
	    calculator.AddOperator(Operator.Mul, multiply);
	    calculator.AddOperator(Operator.Div, divide);

            // 演算方法を指定して演算
	    Console.WriteLine(calculator.Calculate(Operator.Add, 12, 30)); // 出力 : 42
	    Console.WriteLine(calculator.Calculate(Operator.Div, 252, 6)); // 出力 : 42
	}

        private int add(int a, int b) { return a + b; }
	private int subtract(int a, int b) { return a - b; }
	private int multiply(int a, int b) { return a * b; }
	private int divide(int a, int b) { return a / b; }
    }
}
